kong/runloop/handler.lua

```diff
@@ -152,95 +779,274 @@ do
+        end
+      end
+    end
+
+    plugins_version = version
+    plugins = new_plugins
+
+    if version == "init" then
+      if ngx.get_phase() == "init" then
+        log(DEBUG, "initialising plugins done")
+      else
+        log(DEBUG, "initialising plugins on worker #", WORKER_ID, " done")
+      end
+
+    else
+      log(DEBUG, "rebuilding plugins on worker #", WORKER_ID, " done")
+    end
+
+    if recurse then
+      current_version = get_version("plugins")
+      if version ~= current_version then
+        log(DEBUG, "rebuilding plugins on worker #", WORKER_ID)
+        return build_plugins(current_version)
+      end
+    end
+
     return true
   end
 
-
-  -- for unit-testing purposes only
-  _set_check_router_rebuild = function(f)
-    check_router_rebuild = f
+  local function rebuild_sync(callback, version)
+    local pok, ok, err = pcall(callback, version)
+    if not pok or not ok then
+      log(CRIT, "could not rebuild synchronously: ", ok or err)
+    end
   end
 
+  local function rebuild_timer(premature, callback, version, semaphore)
+    if premature then
+      release_semaphore(semaphore)
+      return
+    end
 
-  get_router = function()
-    local version, err = singletons.cache:get("router:version",
-                                              CACHE_ROUTER_OPTS,
-                                              utils.uuid)
-    if err then
-      log(ngx.CRIT, "could not ensure router is up to date: ", err)
-      return nil, err
+    local pok, ok, err = pcall(callback, version, true)
+    if not pok or not ok then
+      log(CRIT, "could not rebuild asynchronously: ", ok or err)
     end
 
-    if version == router_version then
-      return router
+    release_semaphore(semaphore)
+  end
+
+  local function rebuild_async(callback, version, semaphore)
+    local ok, err = ngx.timer.at(0, rebuild_timer, callback, version, semaphore)
+    if not ok then
+      log(CRIT, "could not create rebuild timer: ", err)
+      return false
     end
 
-    -- wrap router rebuilds in a per-worker mutex:
-    -- this prevents dogpiling the database during rebuilds in
-    -- high-concurrency traffic patterns;
-    -- requests that arrive on this process during a router rebuild will be
-    -- queued. once the lock is acquired we re-check the
-    -- router version again to prevent unnecessary subsequent rebuilds
+    return true
+  end
 
-    local opts = {
-      name = "build_router",
-      timeout = build_router_timeout,
-      on_timeout = "run_unlocked",
-    }
-    local ok, err = concurrency.with_coroutine_mutex(opts, check_router_rebuild)
+  local function rebuild(name, callback, version, semaphore, wait)
+    local current_version = get_version(name)
+    if current_version == version then
+      return
+    end
 
+    local ok = acquire_semaphore(semaphore, wait)
     if not ok then
-      return nil, err
+      if wait and wait > 0 then
+        rebuild_sync(callback, current_version, semaphore)
```

Comment:
there are multiple calls to `rebuild_sync` that define a third param `semaphore`, but it does not seem used?

Reply:
@p0pr0ck5 as always, thank you for this! I'll remove that.

Reply:
Fixed now.